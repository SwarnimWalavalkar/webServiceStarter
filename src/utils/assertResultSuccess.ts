import { Result } from "../shared/result";

export function assertResultSuccess<T>(
  result: Result<unknown, Error>
): asserts result is { ok: true; value: T } {
  if (!result.ok) {
    throw result.error;
  }
}
