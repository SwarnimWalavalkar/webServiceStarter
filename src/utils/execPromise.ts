import { exec } from "child_process";

export default function (cmd: string): Promise<string> {
  return new Promise(function (resolve, reject) {
    exec(cmd, function (err, stdout) {
      if (err) return reject(err);
      resolve(stdout.trim());
    });
  });
}
