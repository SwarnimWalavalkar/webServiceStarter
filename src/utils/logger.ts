import pino from "pino";
import als from "../lib/als";

const logger = pino({
  name: "starter-service",
  mixin() {
    return { traceId: als.getStore() };
  },
  level:
    process.env.NODE_ENV === "development"
      ? "debug"
      : process.env.NODE_ENV === "test" || process.env.NODE_ENV === "cli"
      ? "error"
      : "info",
  redact: {
    paths: [
      "password",
      "*.password",
      "*.body.password",
      "event.data.password",
      "req.headers['authorization']",
      "req.cookies['authorization']",
    ],
    censor: "**SENSETIVE INFORMATION**",
  },
});

export default logger;
