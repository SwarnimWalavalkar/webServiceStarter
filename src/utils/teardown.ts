import app from "../app";
import hermes from "../hermes";
import { dbConnection } from "../services/db";
import { redis } from "../services/redis";

export default async function teardown() {
  await app.close();

  await hermes.disconnect();
  await dbConnection.end();
  await redis.quit();
}
