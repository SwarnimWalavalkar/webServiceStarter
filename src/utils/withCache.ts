import { redis } from "../dependencies/redis";

const inMemoryCache = new Map<string, { expiry: number; value: unknown }>();

export default async function withCache<T>(
  computeVal: Promise<T> | (() => Promise<T>),
  key: string,
  ex: number = 60 * 60 * 24
): Promise<T> {
  const inMemoryCacheValue = inMemoryCache.get(key);

  if (inMemoryCacheValue) {
    if (Date.now() > inMemoryCacheValue.expiry) {
      inMemoryCache.delete(key);
    } else {
      return inMemoryCacheValue.value as T;
    }
  }

  const redisCacheValue = await redis.get(key);

  if (redisCacheValue) {
    return JSON.parse(redisCacheValue) as T;
  }

  let computedValue: T;

  if (computeVal instanceof Promise) {
    computedValue = await computeVal;
  } else {
    computedValue = await computeVal();
  }

  if (computedValue) {
    await redis.set(key, JSON.stringify(computedValue), "EX", ex);
    inMemoryCache.set(key, {
      expiry: Date.now() + ex * 1000,
      value: computedValue,
    });
  }

  return computedValue! as T;
}
