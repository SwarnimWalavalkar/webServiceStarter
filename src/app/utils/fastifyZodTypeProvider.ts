import logger from "../../utils/logger";
import { FastifySchemaCompiler } from "fastify";
import { ZodAny, ZodError } from "zod";
import { zodToJsonSchema } from "zod-to-json-schema";
import { fromZodError } from "zod-validation-error";
import { Schema } from "../types/fastify";

const zodToJsonSchemaOptions = {
  target: "openApi3",
  $refStrategy: "none",
} as const;

export const validatorCompiler: FastifySchemaCompiler<ZodAny> =
  ({ schema }) =>
  (data): any => {
    try {
      return { value: schema.parse(data) };
    } catch (error) {
      const validationError = fromZodError(error as ZodError);
      logger.error("[VALIDATION ERROR]", validationError);

      return { error: validationError };
    }
  };

function hasOwnProperty<T, K extends PropertyKey>(
  obj: T,
  prop: K
): obj is T & Record<K, any> {
  return Object.prototype.hasOwnProperty.call(obj, prop);
}

function resolveSchema(
  maybeSchema: ZodAny | { properties: ZodAny }
): Pick<ZodAny, "safeParse"> {
  if (hasOwnProperty(maybeSchema, "safeParse")) {
    return maybeSchema;
  }
  if (hasOwnProperty(maybeSchema, "properties")) {
    return maybeSchema.properties;
  }
  throw new Error(`Invalid schema passed: ${JSON.stringify(maybeSchema)}`);
}

export const createJsonSchemaTransform = ({
  skipList,
}: {
  skipList: readonly string[];
}) => {
  return ({ schema, url }: { schema: Schema; url: string }) => {
    if (!schema) {
      return {
        schema,
        url,
      };
    }

    const { response, headers, querystring, body, params, hide, ...rest } =
      schema;

    const transformed: Record<string, any> = {};

    if (skipList.includes(url) || hide) {
      transformed.hide = true;
      return { schema: transformed, url };
    }

    const zodSchemas: Record<string, any> = {
      headers,
      querystring,
      body,
      params,
    };

    for (const prop in zodSchemas) {
      const zodSchema = zodSchemas[prop];
      if (zodSchema) {
        transformed[prop] = zodToJsonSchema(zodSchema, zodToJsonSchemaOptions);
      }
    }

    if (response) {
      transformed.response = {};

      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      for (const prop in response as any) {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const schema = resolveSchema((response as any)[prop]);

        const transformedResponse = zodToJsonSchema(
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          schema as any,
          zodToJsonSchemaOptions
        );
        transformed.response[prop] = transformedResponse;
      }
    }

    for (const prop in rest) {
      const meta = rest[prop as keyof typeof rest];
      if (meta) {
        transformed[prop] = meta;
      }
    }

    return { schema: transformed, url };
  };
};

const defaultSkipList = [
  "/api/v1/docs/",
  "/api/v1/docs/initOAuth",
  "/api/v1/docs/json",
  "/api/v1/docs/uiConfig",
  "/api/v1/docs/yaml",
  "/api/v1/docs/*",
  "/api/v1/docs/static/*",
];

export const jsonSchemaTransform = createJsonSchemaTransform({
  skipList: defaultSkipList,
});
