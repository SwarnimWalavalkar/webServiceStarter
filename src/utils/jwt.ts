import { createSigner, createVerifier } from "fast-jwt";
import config from "../config";
import { ServiceError } from "../shared/errors";
import logger from "./logger";

const accessTokenSigner = createSigner({
  key: config.jwt.tokenSecret,
  expiresIn: "60m",
});

const refreshTokenSigner = createSigner({
  key: config.jwt.tokenSecret,
  expiresIn: "7d",
});

const verifier = createVerifier({ key: config.jwt.tokenSecret });

export const sign = <T extends Record<PropertyKey, any>>(
  type: keyof typeof config.constants.JWT_TOKEN_TYPES,
  payload: T
) => {
  try {
    switch (type) {
      case "ACCESS_TOKEN":
        return accessTokenSigner(payload);
      case "REFRESH_TOKEN":
        return refreshTokenSigner(payload);
      default:
        throw new ServiceError("JWT Error", "Invalid token type");
    }
  } catch (error) {
    logger.error(error, "[JWT SIGN ERROR]");
    throw error;
  }
};

export const verify = <T>(
  token: string
): (T & { iat: number; exp: number }) | null => {
  try {
    return verifier(token);
  } catch (error: any) {
    if (error.message.includes("expired")) {
      return null;
    }
    logger.error(error, "[JWT VERIFY ERROR]");
    throw error;
  }
};
