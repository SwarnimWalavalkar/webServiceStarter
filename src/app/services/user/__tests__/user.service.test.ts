import { it, describe, expect } from "vitest";
import {
  createUser,
  getUserByUsernameOrEmail,
  updateUser,
} from "../user.service";
import { assertResultSuccess } from "../../../../utils/assertResultSuccess";

describe("User Service", () => {
  it("find users by username", async () => {
    const createdUserRes = await createUser({
      name: "Test User",
      username: "testUsername",
      email: "test@test.com",
      password: "passwordHash",
    });

    expect(createdUserRes.ok).toBe(true);
    assertResultSuccess(createdUserRes);

    const foundUserRes = await getUserByUsernameOrEmail(
      createdUserRes.value.username
    );

    expect(foundUserRes.ok).toBe(true);
    assertResultSuccess(foundUserRes);

    expect(foundUserRes.value).toStrictEqual(createdUserRes.value);
  });

  it("create user", async () => {
    const createdUserRes = await createUser({
      name: "Test User",
      username: "testUsername",
      email: "test@test.com",
      password: "passwordHash",
    });

    expect(createdUserRes.ok).toBe(true);
    assertResultSuccess(createdUserRes);

    const { created_at, updated_at, ...createdUser } = createdUserRes.value;

    expect(createdUser).toMatchSnapshot();
  });

  it("update user", async () => {
    const createdUserRes = await createUser({
      name: "Test User",
      username: "testUsername",
      email: "test@test.com",
      password: "passwordHash",
    });

    expect(createdUserRes.ok).toBe(true);
    assertResultSuccess(createdUserRes);

    const newName = "New Test User";
    const updatedUserRes = await updateUser(createdUserRes.value.email, {
      name: newName,
    });

    expect(updatedUserRes.ok).toBe(true);
    assertResultSuccess(updatedUserRes);

    const foundUserRes = await getUserByUsernameOrEmail(
      createdUserRes.value.username
    );

    expect(foundUserRes.ok).toBe(true);
    assertResultSuccess(foundUserRes);

    expect(foundUserRes.value.name).toBe(newName);
  });
});
