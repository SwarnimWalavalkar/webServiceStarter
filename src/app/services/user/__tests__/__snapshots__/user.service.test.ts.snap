// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`User Service > create user 1`] = `
{
  "email": "test@test.com",
  "id": 4,
  "name": "Test User",
  "password": "passwordHash",
  "roles": [
    "user",
  ],
  "username": "testUsername",
}
`;
